aptos-move/framework/aptos-framework/sources/nonce_validation.move

```diff
@@ -0,0 +1,256 @@
+        let bucket = Bucket {
+            nonces_ordered_by_exp_time: big_ordered_map::new_with_reusable(),
+            nonce_to_exp_time_map: big_ordered_map::new_with_reusable(),
+        };
+
+        if (pre_allocate_slots) {
+            // Initiating big ordered maps with 5 pre-allocated storage slots.
+            // (expiration time, address, nonce) is together 48 bytes.
+            // A 4 KB storage slot can store 80+ such tuples.
+            // The 5 slots should be more than enough for the current use case.
+            bucket.nonces_ordered_by_exp_time.allocate_spare_slots(5);
+            bucket.nonce_to_exp_time_map.allocate_spare_slots(5);
+        };
+        bucket
+    }
+
+    // This method is used to prefill the nonce_table with empty buckets one by one.
+    public entry fun add_nonce_buckets(count: u64) acquires NonceHistory {
+        assert!(exists<NonceHistory>(@aptos_framework), error::invalid_state(E_NONCE_HISTORY_DOES_NOT_EXIST));
+        let nonce_history = &mut NonceHistory[@aptos_framework];
+        for (i in 0..count) {
+            if (nonce_history.next_key <= NUM_BUCKETS) {
+                if (!nonce_history.nonce_table.contains(nonce_history.next_key)) {
+                    nonce_history.nonce_table.add(
+                        nonce_history.next_key,
+                        empty_bucket(true)
+                    );
+                };
+                nonce_history.next_key = nonce_history.next_key + 1;
+            }
+        }
+    }
+
+    // Returns true if the input (address, nonce) pair doesn't exist in the nonce history, and inserted into nonce history successfully.
+    // Returns false if the input (address, nonce) pair already exists in the nonce history.
+    public(friend) fun check_and_insert_nonce(
+        sender_address: address,
+        nonce: u64,
+        txn_expiration_time: u64,
+    ): bool acquires NonceHistory {
+        assert!(exists<NonceHistory>(@aptos_framework), error::invalid_state(E_NONCE_HISTORY_DOES_NOT_EXIST));
+        let nonce_history = &mut NonceHistory[@aptos_framework];
+        let nonce_key = NonceKey {
+            sender_address,
+            nonce,
+        };
+        let bucket_index = sip_hash_from_value(&nonce_key) % NUM_BUCKETS;
+        let current_time = timestamp::now_seconds();
+        if (!nonce_history.nonce_table.contains(bucket_index)) {
+            nonce_history.nonce_table.add(
+                bucket_index,
+                empty_bucket(false)
+            );
+        };
+        let bucket = table::borrow_mut(&mut nonce_history.nonce_table, bucket_index);
+
+        let existing_exp_time = bucket.nonce_to_exp_time_map.get(&nonce_key);
+        if (existing_exp_time.is_some()) {
+            let existing_exp_time = existing_exp_time.extract();
+
+            // If the existing (address, nonce) pair has not expired, return false.
+            if (existing_exp_time >= current_time) {
```

Comment:
this line now seems redundant (because of the line below) and confusing (because we expire based on GC added buffer)

is logic correct if we just remove this?

Reply:
As the current setting of `NONCE_REPLAY_PROTECTION_OVERLAP_INTERVAL_SECS` (65 seconds) is same as the `MAX_EXPIRATION_TIME_FOR_ORDERLESS_TXNS` (65 seconds), the second check supersedes the first check. 

But in the future, if we choose to change the setting for `NONCE_REPLAY_PROTECTION_OVERLAP_INTERVAL_SECS`, then we will have to make sure to add this check back.

Reply:
hm but they are now somewhat required to be the same, with the new logic ? (or at least we require that MAX_EXPIRATION_TIME_FOR_ORDERLESS_TXNS < NONCE_REPLAY_PROTECTION_OVERLAP_INTERVAL_SECS or things will not work

that's also a good reason to additionally perform the check current_time < txn_expiration_time < curren_time + NONCE_REPLAY_PROTECTION_OVERLAP_INTERVAL_SECS

Reply:
> that's also a good reason to additionally perform the check current_time < txn_expiration_time < curren_time + NONCE_REPLAY_PROTECTION_OVERLAP_INTERVAL_SECS

I mean instead of checking ` txn_expiration_time < current_time + MAX_EXPIRATION_TIME_FOR_ORDERLESS_TXNS` in prologue and `txn_expiration_time < current_time + NONCE_REPLAY_PROTECTION_OVERLAP_INTERVAL_SECS` in nonce history, it seems cleaner to just re-introduce the removed check again?